addon.py

```diff
@@ -1,78 +1,135 @@
-                item = xbmcgui.ListItem(file.get('name'), iconImage='DefaultFolder.png')
-                xbmcplugin.addDirectoryItem(handle=addon_handle, url=url, listitem=item, isFolder=True)
+            self.add_item(name=folder['name'],
+                          url=self.build_url(
+                              mode='files',
+                              path=os.path.join(folder['path'], folder['name'])
+                          ),
+                          is_folder=True)
+        xbmcplugin.endOfDirectory(self.handle)
+
+    def play_song(self, path):
+        song_meta_url = urllib.parse.urljoin(self.server, 'db/metadata')
+        response = requests.post(song_meta_url, data={'filepath': path,
+                                                      'token': self.token}).json()
+        song_meta = response['metadata']
+
+        song_url = urllib.parse.urljoin(self.server,
+                                        'media/' + urllib.parse.quote_plus(path))
+        item = xbmcgui.ListItem(
+            path='{}?{}'.format(song_url, urllib.parse.urlencode({'token': self.token}))
+        )
+        item.setInfo('music', {
+            'artist': song_meta['artist'],
+            'album': song_meta['album'],
+            'tracknumber': song_meta['track'],
+            'title': song_meta['title'],
+            'year': song_meta['year']
+        })
+        xbmcplugin.setResolvedUrl(self.handle, succeeded=True, listitem=item)
+
+    def act(self, mode, path):
+        if mode is None:
+            self.build_filesystem()
+        elif mode == 'files':
+            if path is None and len(self.folders) == 1:
+                self.list_folder_contents(self.folders[0])
+            elif path is None and len(self.folders) > 1:
+                self.list_folders(self.folders)
             else:
-                url = build_url({'mode': 'play', 'path': files.get('path') + file.get('name')})
-                item = xbmcgui.ListItem(file.get('name'))
-                item.setProperty('IsPlayable', 'true')
-                xbmcplugin.addDirectoryItem(handle=addon_handle, url=url, listitem=item, isFolder=False)
-        xbmcplugin.endOfDirectory(addon_handle)
-elif mode[0] == 'play':
-    play_song(args.get('path')[0], token=token)
\ No newline at end of file
+                self.list_folder_contents(path)
+        else:
+            self.play_song(path)
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(
+        description='Kodi add-on for mStream server communication'
+    )
+    parser.add_argument('base_url', help='launching URL of the add-on')
+    parser.add_argument('addon_handle', type=int, help='the handle of this add-on')
+    parser.add_argument('-m', '--mode',
+                        help='the operating mode of the add-on',
+                        choices=['files', 'play'])
+    parser.add_argument('-p', '--path',
+                        help='currently browsed path')
+    args = parser.parse_args()
```

Comment:
The whole args thing doesn't work, I think because Kodi is calling the plugins in a diffrent way...

Reply:
Still fails with
```
18:09:53.207 T:6296   ERROR: usage: [-h] base_url addon_handle {files,play} path
18:09:53.207 T:6296   ERROR: : error: argument addon_handle: invalid int value: ''
```
Please set you up a Kodi for testing, I can't do every testing for you, that's not the way how PRs work